tests/test-integration/src/test/java/org/eclipse/jetty/test/RebuildSelectorTest.java

```diff
@@ -0,0 +1,254 @@
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.server.handler.DefaultHandler;
+import org.eclipse.jetty.server.handler.HandlerList;
+import org.eclipse.jetty.servlet.ServletContextHandler;
+import org.eclipse.jetty.util.IO;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.util.thread.ExecutionStrategy;
+import org.eclipse.jetty.util.thread.Scheduler;
+import org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume;
+import org.eclipse.jetty.util.thread.strategy.ProduceExecuteConsume;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.startsWith;
+
+@RunWith(Parameterized.class)
+public class RebuildSelectorTest
+{
+    @Parameterized.Parameters(name = "{0}")
+    public static List<ExecutionStrategy.Factory[]> params()
+    {
+        List<ExecutionStrategy.Factory[]> strategies = new ArrayList<>();
+        strategies.add(new ExecutionStrategy.Factory[]{new ExecuteProduceConsume.Factory()});
+        strategies.add(new ExecutionStrategy.Factory[]{new ProduceExecuteConsume.Factory()});
+        return strategies;
+    }
+
+    private HttpClient client;
+    private Server server;
+
+    @After
+    public void stopServer() throws Exception
+    {
+        server.stop();
+    }
+
+    @Before
+    public void startClient() throws Exception
+    {
+        client = new HttpClient();
+        client.setIdleTimeout(2000);
+        client.setMaxConnectionsPerDestination(1);
+        client.start();
+    }
+
+    @After
+    public void stopClient() throws Exception
+    {
+        client.stop();
+    }
+
+    public RebuildSelectorTest(ExecutionStrategy.Factory strategyFactory) throws Exception
+    {
+        server = new Server();
+
+        CustomServerConnector connector = new CustomServerConnector(server, strategyFactory, 1, 1);
+        connector.setPort(0);
+        server.addConnector(connector);
+
+        ServletContextHandler context = new ServletContextHandler();
+        context.setContextPath("/");
+        context.addServlet(HelloServlet.class, "/hello");
+        context.addServlet(CloseSelectorServlet.class, "/selector/close");
+        context.setAttribute("connector", connector);
+
+        HandlerList handlers = new HandlerList();
+        handlers.addHandler(context);
+        handlers.addHandler(new DefaultHandler());
+
+        server.setHandler(handlers);
+
+        server.start();
+    }
+
+    @Test
+    public void testRebuildServerSelector() throws Exception
+    {
+        // Request /hello
+        assertRequestHello();
+
+        // Request /selector/close
+        assertRequestSelectorClose();
+
+        // Wait for selectors to close from action above
+        TimeUnit.MILLISECONDS.sleep(1000);
```

Comment:
can we do this without a wait.  add a latch to your test selector so  you can see when the restart is finished

Reply:
good idea.